### PR TITLE
doc: document why it's safe against async preemption

### LINKED ISSUE #153: JIT: consider locking Goroutine thread until JIT exits
We could use `runtime.LockOSThread()` to block Goruntime from switching the goroutine via [async preemption](https://github.com/golang/proposal/blob/master/design/24543-non-cooperative-preemption.md) which can happen at any point of execution.

Even though we've tested the JIT execution is stable without locking (as in jit/engine_test.go ), there might be still some condition where the JIT would fail. 

This needs more through investigation on how Goroutine saves CPU registers when switches happens (at source code level), and more test cases to ensure whether or not we need thread blocking via `LockOSThread`.

### PR DESCRIPTION
**Pull request by mathetake**
closes #153 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: internal/engine/compiler/RATIONALE.md [hunk 1]
```diff
@@ -58,6 +58,18 @@ And `nativecall` is actually implemented in [arch_amd64.s](./arch_amd64.s)
...
+
+## Why it's safe to execute runtime-generated machine codes against Goroutine preemption
+
+All the assembly codes are entered via the trampoline implemented as Go Assembler Function (e.g. [arch_amd64.s](./arch_amd64.s)),
+and as of Go 1.20, these assembler functions are considered as _unsafe_ for async preemption:
+- https://github.com/golang/go/blob/go1.20rc1/src/runtime/preempt.go#L406-L407
+- https://github.com/golang/go/blob/9f0234214473dfb785a5ad84a8fc62a6a395cbc3/src/runtime/traceback.go#L227
+
+From the Go runtime point of view, the execution of runtime-generated machine codes is considered as a part of
+that trampoline function. That means they are ensured to be unsafe for async preemption, so there won't be any
```
Comment (codefromthecrypt) [inline internal/engine/compiler/RATIONALE.md:70]: shorten last sentence to "Therefore, runtime-generated machine code is also unsafe for async preemption."

File: internal/engine/compiler/RATIONALE.md [hunk 2]
```diff
@@ -58,6 +58,18 @@ And `nativecall` is actually implemented in [arch_amd64.s](./arch_amd64.s)
 as a convenience layer to comply with the Go's official calling convention.
 We delegate the task to jump into the code segment to the Go assembler code.
 
+
+## Why it's safe to execute runtime-generated machine codes against Goroutine preemption
+
```
Comment (codefromthecrypt) [inline internal/engine/compiler/RATIONALE.md:63]: there is a little jargon, so possibly an overview sentence or two can help.

Maybe start with the problem of goroutine premption, and then say "unsafe" means goroutines are not allowed to pre-empt.

**Review by codefromthecrypt — COMMENTED** [review]
Thanks a lot!

File: internal/engine/compiler/RATIONALE.md (hunk 1 above)
Comment (mathetake) [reply internal/engine/compiler/RATIONALE.md:70]: `ensured to be unsafe` is the important part and means `correctly marked as unsafe`, so will modify it.

**Review by codefromthecrypt — APPROVED** [review]
nicely done!

**Pull request merged**